holoviews/plotting/bokeh/callbacks.py

```diff
@@ -1,446 +1,225 @@
-from bokeh.core.json_encoder import serialize_json
-from bokeh.io import _CommsHandle
-from bokeh.util.notebook import get_comms
+from ...streams import (Stream, PositionXY, RangeXY, Selection1D, RangeX,
+                        RangeY, PositionX, PositionY, Bounds)
+from ..comms import JupyterCommJS
 
 
-class Callback(param.ParameterizedFunction):
+def attributes_js(attributes):
     """
-    Callback functions provide an easy way to interactively modify
-    the plot based on changes to the bokeh axis ranges, data sources,
-    tools or widgets.
-
-    The data sent to the Python callback from javascript is defined
-    by the plot attributes and callback_obj attributes and optionally
-    custom javascript code.
-
-    The user should define any plot_attributes and cb_attributes
-    he wants access to, which will be supplied in the form of a
-    dictionary to the call method. The call method can then apply
-    any processing depending on the callback data and return the
-    modified bokeh plot objects.
+    Generates JS code to look up attributes on JS objects from
+    an attributes specification dictionary.
+    """
+    code = ''
+    for key, attr_path in attributes.items():
+        data_assign = "data['{key}'] = ".format(key=key)
+        attrs = attr_path.split('.')
+        obj_name = attrs[0]
+        attr_getters = ''.join(["['{attr}']".format(attr=attr)
+                                for attr in attrs[1:]])
+        code += ''.join([data_assign, obj_name, attr_getters, ';\n'])
+    return code
+
+
+class Callback(param.Parameterized):
+    """
+    Provides a baseclass to define callbacks, which return data from
+    bokeh models such as the plot ranges or various tools. The callback
+    then makes this data available to any streams attached to it.
+
+    The defintion of a callback consists of a number of components:
+
+    * handles    :  The handles define which object the callback will be
+                    attached on.
+
+    * attributes :  The attributes define which attributes to send back
+                    to Python. They are defined as a dictionary mapping
+                    between the name under which the variable is made
+                    available to Python and the specification of the
+                    attribute. The specification should start with the
+                    variable name that is to be accessed and the
+                    location of the attribute separated by periods.
+                    All plotting handles such as tools, the x_range,
+                    y_range and (data)source can be addressed in this
+                    way.
+    * code        : Defines any additional JS code to be executed,
+                    which can modify the data object that is sent to
+                    the backend.
+
+    The callback can also define a _process_msg method, which can
+    modify the data sent by the callback before it is passed to the
+    streams.
     """
```

Comment:
> The user can supply some custom code to the instance to be executed alongside the callback

Sure! In that case you'll need to call ``super`` :-)

In the end, I do think class attributes are fine now I see how they are defined on each concrete Callback...